spatialos-sdk/src/worker/entity.rs

```diff
@@ -26,114 +25,70 @@ impl Entity {
-    ) -> Result<(), String> {
-        let vtable = DATABASE.get_vtable(component_id).unwrap();
-        let deserialize_func = vtable
-            .worker_vtable
-            .component_data_deserialize
-            .unwrap_or_else(|| {
-                panic!(
-                    "No component_data_deserialize method defined for {}",
-                    component_id
-                )
-            });
-
-        // Create the **void that the C API requires. We need to then clean this up later.
-        // The value pointed to by handle_out_ptr is written to during the deserialize method.
-        let placeholder_ptr = Box::into_raw(Box::new(0)) as *mut ::std::os::raw::c_void;
-        let handle_out_ptr = Box::into_raw(Box::new(placeholder_ptr));
-
-        let deserialize_result = deserialize_func(
-            component_id,
-            ptr::null_mut(),
-            component.as_ptr_mut(),
-            handle_out_ptr,
-        );
-
-        match deserialize_result {
-            1 => {},
-            0 => return Err("Error deserializing serialized data. Is the SchemaComponentData malformed?".to_owned()),
-            _ => panic!("Unexpected return value from deserialize function. Expected true or false. Received other.")
-        };
-
-        let component_data = Worker_ComponentData {
-            reserved: ptr::null_mut(),
-            component_id,
-            schema_type: ptr::null_mut(),
-            user_handle: *handle_out_ptr,
-        };
-
-        // Reconstruct these objects so they can be de-alloc'ed. We have pulled the data required
-        // from them by de-referencing handle_out_ptr above.
-        Box::from_raw(placeholder_ptr);
-        Box::from_raw(handle_out_ptr);
-
-        self.add_raw(&component_data)
+    pub fn get<C: Component>(&self) -> Option<schema::Result<C>> {
+        self.components.get(&C::ID).map(|data| data.deserialize())
     }
 
-    pub fn get<C: Component>(&self) -> Option<&C> {
+    /// Converts the entity's contents into a format that can be used with the C API.
+    ///
+    /// In cases where the C API takes an entity as a parameter, it does so by taking an array of
+    /// `Worker_ComponentData` structs. This method gathers the serialized component data and
+    /// returns a `Vec<Worker_ComponentData>` that can be used to pass data to the C API.
+    ///
+    /// # Safety
+    ///
+    /// * The returned `Worker_ComponentData` objects borrow the serialized schema data from the
+    ///   `Entity`, and so cannot be used safely once the `Entity` is dropped.
+    /// * The returned `Worker_ComponentData` objects should only be used with C API functions that
+    ///   will not mutate the objects pointed to by the `schema_type` fields. In particular, this
+    ///   should be safe to use with `Worker_Connection_SendCreateEntityRequest` and
+    ///   `Worker_SnapshotOutputStream_WriteEntity`.
+    pub(crate) fn raw_component_data(&self) -> Vec<Worker_ComponentData> {
```

Comment:
Actually, you'll probably want to make this function bind to `self` rather than `&self` and relinquish ownership of the data, as that is the documented semantics for `Schema_*` objects (which is actually different to vtable based user handles - sorry).

Reply:
To preserve the existing behaviour, you could also add a `clone` function to `Entity` if it doesn't exist already after doing the change above, then change:
```
entity.raw_component_data()
```
to
```
entity.clone().raw_component_data()
```
in places where we call the C API.

Reply:
Alternatively, `raw_component_data()` could deep copy the schema data since there are built in schema functions for those sorts of things

Reply:
I like the plan of changing it to take `self` and implement `Clone` (which I _think_ I already did). That feels more idiomatic to me, in that you only pay the cost of deep-copying the data if you need to.

Reply:
I've renamed the function to `into_raw()` and updated the comments to reflect the nuances of the new API 😎